source/base/hdf5.cc

```diff
@@ -0,0 +1,1513 @@
+    // the case of std::complex types the HDF5 handlers are automatically freed
+    // using the destructor of std::shared_ptr.
+    template <typename number>
+    std::shared_ptr<hid_t>
+    get_hdf5_datatype()
+    {
+      std::shared_ptr<hid_t> t_type;
+      if (std::is_same<number, float>::value)
+        {
+          t_type  = std::shared_ptr<hid_t>(new hid_t);
+          *t_type = H5T_NATIVE_FLOAT;
+        }
+      else if (std::is_same<number, double>::value)
+        {
+          t_type  = std::shared_ptr<hid_t>(new hid_t);
+          *t_type = H5T_NATIVE_DOUBLE;
+        }
+      else if (std::is_same<number, int>::value)
+        {
+          t_type  = std::shared_ptr<hid_t>(new hid_t);
+          *t_type = H5T_NATIVE_INT;
+        }
+      else if (std::is_same<number, unsigned int>::value)
+        {
+          t_type  = std::shared_ptr<hid_t>(new hid_t);
+          *t_type = H5T_NATIVE_UINT;
+        }
+      else if (std::is_same<number, std::complex<float>>::value)
+        {
+          t_type  = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+            // Relase the HDF5 resource
+            H5Tclose(*pointer);
+            delete pointer;
+          });
+          *t_type = H5Tcreate(H5T_COMPOUND, sizeof(std::complex<float>));
+          //  The C++ standards committee agreed to mandate that the storage
+          //  format used for the std::complex type be binary-compatible with
+          //  the C99 type, i.e. an array T[2] with consecutive real [0] and
+          //  imaginary [1] parts.
+          H5Tinsert(*t_type, "r", 0, H5T_NATIVE_FLOAT);
+          H5Tinsert(*t_type, "i", sizeof(float), H5T_NATIVE_FLOAT);
+        }
+      else if (std::is_same<number, std::complex<double>>::value)
+        {
+          t_type  = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+            // Relase the HDF5 resource
+            H5Tclose(*pointer);
+            delete pointer;
+          });
+          *t_type = H5Tcreate(H5T_COMPOUND, sizeof(std::complex<double>));
+          //  The C++ standards committee agreed to mandate that the storage
+          //  format used for the std::complex type be binary-compatible with
+          //  the C99 type, i.e. an array T[2] with consecutive real [0] and
+          //  imaginary [1] parts.
+          H5Tinsert(*t_type, "r", 0, H5T_NATIVE_DOUBLE);
+          H5Tinsert(*t_type, "i", sizeof(double), H5T_NATIVE_DOUBLE);
+        }
+      else
+        {
+          Assert(false, ExcInternalError());
+        }
+      return t_type;
+    }
+
+    // This function returns the pointer to the raw data of a container
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, std::vector<number>>::value,
+      std::vector<hsize_t>>::type
+    get_container_dimensions(const Container<number> &data)
+    {
+      return std::vector<hsize_t>{data.size()};
```

Comment:
I added the following comment:

Several HDF5 functions such as H5Screate_simple() require a one-dimensional array that specifies the size of each dimension of the container, see: https://support.hdfgroup.org/HDF5/doc1.8/RM/RM_H5S.html#Dataspace-CreateSimple
The function get_container_dimensions returns a vector with the dimensions of the container.
For a std::vector the function returns std::vector<hsize_t>{vector_size}
For a Vector the function returns std::vector<hsize_t>{vector_size}
For a FullMatrix the function returns std::vector<hsize_t>{rows, columns}

Instead of using functions with std::enable_if, "constexpr if" from C++17 could be used

Reply:
What I mean is this: I know that if you call
```
  std::vector<int>(3);
```
that this creates a vector of size 3, initialized with zeros. And that if you call
```
  std::vector<int>{0, 12, 42, 288};
```
it will create a vector of size 4 with the given elements. But
```
  std::vector<int>{3};
```
that case is ambiguous at least to my eyes (though not to the compiler): Is it creating a list of length 3, or a list of length 1 with one element equal to 3? If *I* can't remember what this syntax does, others will likely not be able to either :-)

So I would prefer if here and below you'd disambiguate this. I *think* what you're doing is return a list of length 1 with the single element equal to `data.size()`.